## Building the Transcriptome 

I used a combination of building my own transcriptome using **[Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)** to compare gene content of _P. bifurca_ to _P. picta_. I also aligned my RNA data to the _P. picta_ transcriptome reference as it is more complete with annotations and to be able to compare between species. Python scripts can be found **[here](https://github.com/ljmfong/Poecilia-bifurca-Characterizing-Sex-Chromosome/tree/main/python_scripts/Trinity_scripts)** and all Rscripts *[here]()**.

------------------------------------------------------------------------------------------------------------------------------------
#### a. Checking RNA quality 
**[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)**: 
Make a text file with the pathway & name to all the files you want to QC (see bifurca_rna_samples.txt for an example - ensure there is a space after each line so FASTQC knows to read each file independently): Make an output folder for your fastqcs:

    FILE="bifurca_rna_samples.txt"; for i in $(cat "$FILE"); do /Linux/bin/fastqc $i -o rna_fastqc_untrimmed; done

Example if you want to only do one file:

    /Linux/bin/fastqc rna_seq/bif_Female_1_RNA_S23_L002_R1_001.fastq.gz -o rna_fastqc_untrimmed/

Check multiple fastqc using multiqc:
**[MultiQC](https://multiqc.info/)**: Merge multiple FastQC files. Ensure that all the fastqcs are in the same directory.


    cd rna_fastqc_untrimmed/
    /Linux/bin/multiqc ./

* Important QC reports include: Per base sequence quality, Per sequence quality scores, Per seq GC content, Overrepresented sequences

#### b. RNA Trimming

**[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)**: Trim off adaptors. Example command:

Using the script to trim multiple files at once: 

    /Linux/bin/python ~/bin/01.trimmomatic_bifurca_RNA.py ~/bifurca_project/rna_seq/ /Linux/Trimmomatic-0.36/trimmomatic-0.36.jar /home/ljmfong/bifurca_project/universal_adapter.fa >trimmomatic_universal.txt

Running just trimmomatic for one file at a time:

    java -jar /Linux/Trimmomatic-0.36/trimmomatic-0.36.jar PE -phred33 sample_1_R1.fastq.gz sample_1_R2.fastq.gz sample_1_output_forward_paired.fq.gz sample_1_output_forward_unpaired.fq.gz sample_1_output_reverse_paired.fq.gz sample_1_output_reverse_unpaired.fq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:50

* Run FastQC on the trimmed sequences and compare results.
    
------------------------------------------------------------------------------------------------------------------------------------
### Building the Transcriptome

#### a. Alignment with HISAT2:
We used **[HISAT2](https://daehwankimlab.github.io/hisat2/)** for our alignment to first see how the trimmed RNA sequences aligned to the draft genome. You will first need to build an index, then align - Example comamand:

     /Linux/bin/hisat2-build -f 05.ragtag_output/ragtag_scaffolds_only.fasta 05.ragtag_output/bifurca_ragtag_scaffolds_only 
     /Linux/bin/hisat2 05.ragtag_output/bifurca_ragtag_scaffolds_only -1 sample_R1_001.output_forward_paired.fq.gz -2 sample_R2_001.output_reverse_paired.fq.gz --phred33 -q -p 12 --no-discordant --no-mixed --no-unal --dta -S sample_hisat2.sam --met-file sample_hisat2.met --summary-file sample_hisat2_summary.txt

#### b. Constructring _de novo_ assembly

**[Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)**: Build _de novo_ assembly. Trinity has a three step-process (Inchworm, Chrysalis, and Butterfly), some of the steps can take a long time. It is recommended to build the assembly in steps.

* Make a tab-delimited file that includes all trimmed forward and reverse sample file names (together with their paths). This will ensure that all samples are used to create the assembly. File should be of the format: 
    sample_name \t sample_forward.fq.gz \t sample_reverse.fq.gz 

Then run:
##### i. Inchworm step:

    trinity2.11.0/Trinity --seqType fq --samples_file picta_reads.txt --CPU 32 --normalize_max_read_cov 50 --max_memory 1500G --no_run_chrysalis >trinity_normalization_inchworm.txt
      
##### ii. Chrysalis step:

      trinity2.11.0/Trinity --seqType fq --samples_file picta_reads.txt --CPU 32 --normalize_max_read_cov 50 --max_memory 1500G --no_distributed_trinity_exec >trinity_chrysalis.txt
  
##### iii. Butterfly step:

      trinity2.11.0/Trinity  --seqType fq --samples_file picta_reads.txt --CPU 32 --max_memory 1500G >trinity_full.txt


Note the Trinity output format. You will get a '**[Trinity.fasta](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Output-of-Trinity-Assembly)**' output in a new 'trinity_out_dir/' that is created when Trinity is run. Trinity clusters transcripts loosely into 'genes' and contains 'isoforms' of those genes. Custom script is used to determine assembly statistics (**03.assembly_stats.py**). Script takes the trinity output folder, searches for the assembly fasta file and calculates basic stats.

    python 03.assembly_stats.py trinity_out_dir/



#### c. Filtering Transcriptome 
Steps to filter transcriptome, removing redundancy,non-coding RNA and transcripts without an open reading frame

**[RSEM](https://deweylab.github.io/RSEM/)** & **[Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) Using Trinity Scripts**: Map reads to the _de novo_ transcriptome. 
Trinity has scripts that measure gene expression (RSEM) of reads and aligns them to the _de novo_ assembly (Bowtie2). These scripts can be found in the 'util' directory when Trinity is installed. However, both softwards must be installed and their PATH must be set in order to run the Trinity scripts. Run the following scripts:
* Prepare reference for alignment and abundance estimation:

        trinity2.11.0/util/align_and_estimate_abundance.pl --transcripts trinity_out_dir/Trinity.fasta --est_method RSEM --aln_method bowtie2 --trinity_mode --prep_reference >prep_rsem_reference.txt
        
* Run alignment and abundance esetimation: Example command:

        trinity2.11.0/util/align_and_estimate_abundance.pl --transcripts trinity_out_dir/Trinity.fasta --seqType fq --left suriname_picta/Trimmed_Reads/T1-Dame_R1_098.output_forward_paired_50bp.fastq.gz --right suriname_picta/Trimmed_Reads/T1-Dame_R2_098.output_reverse_paired_50bp.fastq.gz --est_method RSEM --aln_method bowtie2 --trinity_mode --output_dir picta_rsem/T1-Dame/ --bowtie2_RSEM '--no-mixed --no-discordant --gbar 1000 --end-to-end -k 200 --phred33-quals' >T1-Dame-trinity-rsem.txt 2>T1-Dame-trinity-rsem-alignment.txt


Custom scripts are used to filter transcriptome to remove redundancy (**04.get-best-isoform.py** and **05.get-best-isoform-fasta.py**):

        python 04.get-best-isoform.py picta_rsem/ trinity_out_dir/Trinity.fasta best_isoform.txt >04_best_isoform.txt
        python 05.get-best-isoform-fasta.py trinity_out_dir/Trinity.fasta best_isoform.txt >05_get_best_isoform.txt

It will save in the trinity_out_dir folder as Trinity.bestisoform.fa 

**[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)**: Identify non-coding RNA (ncRNA): Download fasta file of ncRNA from closely related species from **[ENSEMBL](http://uswest.ensembl.org/index.html)** e.g. Oryzias_latipes.MEDAKA1.ncrna.fa. If using BLAST on command line, **[BLAST Commandline Manual](https://www.ncbi.nlm.nih.gov/books/NBK279690/)**.

* Build BLAST index:


        makeblastdb -in Oryzias_latipes.ASM223467v1.ncrna.fa -input_type fasta -dbtype nucl -title MEDAKA.ncrna_db -out MEDAKA.ncrna_db >make_blast.txt

* BLAST transcript to ncRNA of related species:

        blastn -evalue 10e-10 -db MEDAKA.ncrna_db -query trinity_out_dir/Trinity.bestisoform.fa -out trinity_bestisoform_fasta_file_MEDAKAncrna.blastout -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq"

Custom scripts are used to identify top BLAST hit (**06.get-ncrna.py**) and filter the assembly (**07.filter-assembly-ncrna.py**):

        python 06.get-ncrna.py trinity_bestisoform_fasta_file_MEDAKAncrna.blastout >06_id_blast_tophits.txt
Create a new folder and put the output of the previous command (.tophits) into this folder.

        python 07.filter-assembly-ncrna.py picta_blast_tophits/ trinity_out_dir/Trinity.bestisoform.fa >07_filter_assembly.txt

This output will appear in the trinity_out_dir folder.

**[Transdecoder](https://github.com/TransDecoder/TransDecoder/wiki)**: Remove transcripts without open reading frames (ORFs).
* Extract long ORFs:

        TransDecoder/TransDecoder.LongOrfs -t trinity_out_dir/Trinity.bestisoform_ncrnafiltered.fa -m 150 >extract_orfs.txt

* Final coding region predictions:

        TransDecoder/TransDecoder.Predict -t trinity_out_dir/Trinity.bestisoform_ncrnafiltered.fa --retain_long_orfs_length 150 >codingregion_predicts.txt 2>codingregion_predicts_2.txt


**[CAP3](http://seq.cs.iastate.edu/cap3.html)**: Assemble contigs.

        CAP3/cap3 Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cds -s 900 >cap3_assembly.txt
        
* Merge CAP3 singlets and contigs:

        cat Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cds.cap.contigs Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cds.cap.singlets >Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa


**[BWA](http://bio-bwa.sourceforge.net/)**: Map trimmed reads to the final assembly
* Build BWA index:

        bwa index Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa >bwa_index.txt
        
* Align reads: Example Command:

        bwa aln -t 12 Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa suriname_picta/Trimmed_Reads/T1-Dame_R1_098.output_forward_paired_50bp.fastq.gz >T1-Dame_forward.sai 2>T1-Dame_forward.txt

* Generate SAM format: Example Command:

        bwa sampe -r "@RG\tID:sample_name" Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa sample_forward.sai sample_reverse.sai sample_forward_paired.fastq.gz sample_reverse_paired.fastq.gz >sample_sampe.sam 2>sample_sampe.txt


**[SAMtools](http://www.htslib.org/doc/samtools.html)**: Convert SAMs to BAMs
* Build index:

        samtools faidx Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa
        
* Compress into BAM files: Example Command:

        samtools view -t Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa.fai -F 4 -h -S -b -o sample_sampe.bam sample_sampe.sam
        
* Order individual BAM files: Example Command:

        samtools sort -o sample.sorted sample_sampe.bam


#### d. Getting count data:
We used **[HTSeq](https://htseq.readthedocs.io/en/release_0.11.1/count.html)** to get gene expression data, given the alignment from the RNAseq data to the Trinity assembly. Get the count data then extract the counts:

    mkdir HTSeq_counts
    /Linux/bin/htseq-count -f bam -r pos -s no sample.sorted Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap_nogaps.gtf > HTSeq_counts/sample_sorted.sam_htseqcount.txt
    cd HTSeq_counts
    
    for f in *.sam_htseqcount.txt
    do
      subdir=${f%%.*}
      [ ! -d "$subdir" ] && mkdir -- "$subdir"
      mv -- "$f" "$subdir"
    done

    mkdir count_extractions
    /Linux/bin/python 11.extract-counts.py HTSeq_counts extract_all_read_counts.txt
    /Linux/bin/python L12.extract-counts-annotated.py extract_all_read_counts.txt Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap_nogaps.gtf count_extractions/bifurca_gene_position.txt
    /Linux/bin/python 13.extract-gene-length.py extract_all_read_counts.txt Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap_nogaps.gtf count_extractions/bifurca_gene_length.txt

Use the convert-to-RPKM_allspp.R to then convert the count data to RPKM.


