## Characterizing the Sex Chromosome in _P. bifurca_

There are many different analyses that can be used to identify sex chromosomes. Here, we use a combination of male to female coverage analysis, single-nucleotide polymorphisms (SNP) density, and k-mer analysis. Python scripts can be found **[here](https://github.com/ljmfong/Poecilia-bifurca-Characterizing-Sex-Chromosome/tree/main/python_scripts)** and all Rscripts **[here]()**.

------------------------------------------------------------------------------------------------------------------------------------

###  a. Coverage Analysis

Following the _de novo_ genome assembly, we first aligned all the short-read DNA-sequence data to the _de novo_ genome assembly using **[BWA](https://bio-bwa.sourceforge.net/bwa.shtml)**. The following is an example command, remember to build a BWA index. Remember for paired-end data to do it for both forward and reverse sequences:

    bwa index ~/bifurca_project/ragtag_output/ragtag_scaffolds_only.fasta
    bwa aln ~/bifurca_project/ragtag_output/ragtag_scaffolds_only.fasta ~/bifurca_project/dna_concatenated_paired_reads/bif_female_1.output_forward_paired.fq.gz > Female_1_forward.sai
    bwa aln ~/bifurca_project/ragtag_output/ragtag_scaffolds_only.fasta ~/bifurca_project/dna_concatenated_paired_reads/bif_female_1.output_reverse_paired.fq.gz > Female_1_reverse.sai

Create a SAM file using bwa and find _unique_ mapping reads, the following is an example command:

    bwa sampe ~/bifurca_project/ragtag_output/ragtag_scaffolds_only.fasta ~/bifurca_project/bwa_cov_analysis/Female_1_forward.sai ~/bifurca_project/bwa_cov_analysis/Female_1_reverse.sai ~/bifurca_project/dna_concatenated_paired_reads/bif_female_1.output_forward_paired.fq.gz ~/bifurca_project/dna_concatenated_paired_reads/bif_female_1.output_reverse_paired.fq.gz > Female_1_sampe.sam
    grep 'XT:A:U' Female_1_sampe.sam > Female_1_uniq.sam

You can then estimate coverage with SOAP.coverage (see attached script). The following is an example command:

    bin/soap.coverage -sam -cvg -refsingle ~/bifurca_project/ragtag_output/ragtag_scaffolds_only.fasta -i ~/bifurca_project/bwa_sampe/Female_1_uniq.sam -o ~/bifurca_project/soapcoverage_uniq/Female_1_soapCov.txt -p 12 -window Female_1_windows.txt 50000

You can extract the coverage using the 06.extract_coverage.py script. NOTE: You need to have the soapCov.txt files in two separate directories - one for females and one for males. Also NOTE: The CovList shows the coverage of the inputted individuals (i.e. there were will be 3 values separated by "|" because I had 3 females and 3 males. In the python script, for line 28 - make sure the files in the folder match what the file name ends with - e.g. "windows.txt" or "extractcov.txt".


    python 06.extract_coverage.py ~/bifurca_project/Female_soapcov/ ~/bifurca_project/coverage_analysis/female_coverage_only.txt
    python 06.extract_coverage.py ~/bifurca_project/Male_soapcov/ ~/bifurca_project/coverage_analysis/male_coverage_only.txt

You can then manually calculate the fold change in Excel or R, and plot using the according Rscript.

------------------------------------------------------------------------------------------------------------------------------------

###  b. SNP Analysis

First we have to map using **[Bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml)** the bifurca FASTq to the bifurca draft genome assmebly then sort usng SAMTools. Build an index and then run the alignment. Because Bowtie2 can't read zipped files, you may need to unzip your DNA files - example commands:

    /Linux/bin/bowtie2-build -f 05.ragtag_output/ragtag_scaffolds_only.fasta 05.ragtag_output/ragtag_scaffolds_only.bowtie_index > ragtag_scaffolds_only.bowtie_index_run.txt    
    gunzip bif_female_1.output_forward_paired.fq.gz
    gunzip bif_female_1.output_reverse_paired.fq.gz
    /Linux/bin/bowtie2 -q --phred33 -p 12 -x 05.ragtag_output/ragtag_scaffolds_only.bowtie_index -1 bif_female_1.output_forward_paired.fq -2 bif_female_1.output_reverse_paired.fq -S bif_female_1_bowtie.map 2> bif_female_1_bowtie_map_stats.txt
    /Linux/bin/samtools sort bif_female_1_bowtie.map -o bif_female_1_bowtie.map.sorted

Converting the map.sorted files into profiles requires **[sam2pro](http://guanine.evolbio.mpg.de/mlRho/)** and you can inspect the profiles:

    /Linux/bin/samtools mpileup bif_female_1_bowtie.map.sorted | ~/bin/Sam2pro_0.8/sam2pro -c 5 > bif_female_1_bowtie.map.sorted.pro
    ~/bin/formatPro_0.5/FormatPro_0.5/formatPro bif_female_1_bowtie.map.sorted.pro #This should result in a profileDb.con, profileDb.pos, and profileDb.sum file
    
    ~/bin/inspectPro_0.3/InspectPro_0.3/inspectPro bif_female_1_profileDb.con | head #Contigs are listed in this file:
    ~/bin/inspectPro_0.3/InspectPro_0.3/inspectPro bif_female_1_profileDb.con | tail -n +2 | sort -k 2 -n -r | head #To find that the longest contig, which consists of 48,930,977 positions with a coverage >= 4
    ~/bin/inspectPro_0.3/InspectPro_0.3/inspectPro bif_female_1_profileDb.pos | head #To find position 49 is occupied by profile 0, position 50 by profile 1
    ~/bin/inspectPro_0.3/InspectPro_0.3/inspectPro ~/bifurca_project/08.SNP_density/formatted_profiles/bif_female_1_profileDb.sum | head #This means that profile 0 was found 856,448 times and consists of 4 Ts, and so on.
    ~/bin/inspectPro_0.3/InspectPro_0.3/inspectPro bif_female_1_profileDb.sum | tail -n +2 | sort -n -k 2 -r | head #This will tell you which profile was found most frequently, type
    
Get the names (genes) with the SNP information - see **[Building the Transcriptome.md](https://github.com/ljmfong/Poecilia-bifurca-Characterizing-Sex-Chromosome/blob/main/Building%20the%20Transcriptome.md)** for more information. You will need to get the coverage threshold:

    /Linux/bin/python 08.create_coord_file.py count_extractions/bifurca_gene_position.txt Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap_nogaps.gtf gene_coordinates.txt
    grep ">" bif_female_1_bowtie.map.sorted.pro  > bif_female_1_bowtie.map.sorted.pro.list
    cat bif_female_1_bowtie.map.sorted.pro | awk '(($2+$3+$4+$5)>10 || /Scaffold/ )' bif_female_1_bowtie.map.sorted.pro.sites10
    cat bif_female_1_bowtie.map.sorted.pro.sites10 | awk '(($2+$3+$4+$5)>10 || /Scaffold/ )' | awk '( /Scaffold/ || ($2>(0.3*($2+$3+$4+$5)) && ($3>(0.3*($2+$3+$4+$5)))) || ($2>(0.3*($2+$3+$4+$5)) && ($4>(0.3*($2+$3+$4+$5)))) || ($2>(0.3*($2+$3+$4+$5)) && ($5>(0.3*($2+$3+$4+$5)))) || ($3>(0.3*($2+$3+$4+$5)) && ($4>(0.3*($2+$3+$4+$5)))) || ($3>(0.3*($2+$3+$4+$5)) && ($5>(0.3*($2+$3+$4+$5)))) || ($4>(0.3*($2+$3+$4+$5)) && ($5>(0.3*($2+$3+$4+$5))))    )' > bif_female_1_bowtie.map.sorted.pro.sites10.SNP30
    sed -i 's/[>]//g' bif_female_1_bowtie.map.sorted.pro.sites10.SNP30
    sed -i 's/[>]//g' bif_female_1_bowtie.map.sorted.pro.sites10
   
Everything needs to be a path (directories) in the following script and samples should also be in their own directories, e.g.:
Move bif_female_1_bowtie.map.sorted.pro.list to scaffold_names; Move bif_female_1_bowtie.map.sorted.pro.sites10 to cov_thresh; move bif_male_3_bowtie.map.sorted.pro.sites10.SNP30 to maf

    mkdir scaffold_names
    mkdir cov_thresh
    mkdir maf
    mkdir extract_snp_genes
    /Linux/bin/python 09.extract_SNPs_genes_sampleseparate.py scaffold_names/ cov_thresh/ maf/ gene_coordinates.txt extract_snp_genes/
    mkdir extract_snp_genes/males 
    mkdir extract_snp_genes/females
    # Move the outputted files into the respetive directories
    mkdir extract_snp_genes/males_bygene_serial
    mkdir extract_snp_genes/females_bygene_serial
    /Linux/bin/python 2.extract_SNPdensity.1.bygene.serial.py extract_snp_genes/males extract_snp_genes/males_bygene_serial
    
For convenience, you can also rename the Scaffolds to the respective Linkage Group (LG) numbers then calculate fold change:

    /Linux/bin/python L00.convert_scaffold_toLG.py  count_extractions/bifurca_gene_position.txt scaffold_to_LG.txt
    mkdir foldchange/
    /Linux/bin/python 11.bifurca_calculate_foldchange.bygene.py extract_snp_genes/females_bygene_serial/Femalegenome.males.genesgtffiltered.1.bygene extract_snp_genes/males_bygene_serial/Femalegenome.males.genesgtffiltered.1.bygene foldchange/ scaffold_to_LG.txt

Use the Rscript to make a density plot    


------------------------------------------------------------------------------------------------------------------------------------

###  c. Kmer Analysis

Find out how many basepairs your females and male reads contained:

First concatenate your reads if you haven't done so, example command:

    cat bif_female_1.output_forward_paired.fq.gz bif_female_1.output_reverse_paired.fq.gz >bif_female_1_both.fq.gz
    cat bif_female_1_both.fq.gz bif_female_2_both.fq.gz bif_female_3_both.fq.gz > all_bif_female_DNA.fq.gz


Then run **[Jellyfish](https://github.com/gmarcais/Jellyfish)** to get a histogram to know the profile (your expectations):

    /Linux/bin/jellyfish-2.3.3 count -C -m 31 -s 100M -t 12 <(zcat all_bif_female_DNA.fq.gz)
    /Linux/bin/jellyfish-2.3.3 count -C -m 31 -s 100M -t 12 <(zcat all_bif_male_DNA.fq.gz)

    /Linux/bin/jellyfish-2.3.3 histo ~/bifurca_project/female_mer_counts.jf > all_female_bif_counts.histo
    /Linux/bin/jellyfish-2.3.3 histo ~/bifurca_project/male_mer_counts.jf > all_male_bif_counts.histo


You can further do a k-mer comparison between species following **[HAWK](https://github.com/atifrahman/HAWK)** 

    bash countKmers 2>countKmers.log
    bash runHawk 2>runHawk_firstrun.log
    /Linux/HAWK-1.7.0-beta/supplements/EIG6.0.1-Hawk/bin/smartpca -p /Linux/HAWK-1.7.0-beta/parfile.txt
    bash post_runHawk 2>post_runHawk.log

* Edit the scripts accordingly (e.g. location and name of your files) since they are BASH scripts
* Organize the sorted_files.txt, total_kmer_counts.txt, and gwas_info.txt to represent each family and save them into each family's individual directories (runHawk is in the same directory that has the "sorted_files.txt" and "total_kmer_counts.txt"):
You will also have to make the manual case and control total_kmer_count files


At times, HAWK may not always properly produce the sums. To ensure that they have done so correctly, use the awk command to make sure they have summed correctly - example command:

    
    awk '{print $1,$5,$6,$7}' case_out_w_bonf_sorted.kmerDiff | awk '{print $1; for(i=1; i<=NF;i++) j+=$i; print j; j=0 }'  | sed 'N;s/\n/\t/' > sum_case_out_w_bonf_sorted.kmerDiff
    awk '{print $1,$8,$9,$10}' case_out_w_bonf_sorted.kmerDiff | awk '{print $1; for(i=1; i<=NF;i++) j+=$i; print j; j=0 }'  | sed 'N;s/\n/\t/' > sum_case_control_out_w_bonf_sorted.kmerDiff
    cut -f2 sum_case_control_out_w_bonf_sorted.kmerDiff | sum_case_out_w_bonf_sorted.kmerDiff - > corrected_sums_case_out_w_bonf_sorted.txt

* Note that CASE = Males, and CONTROL = Females. The header of .kmerDiff files looks like the following:

    Kmer-seq	sum_case_kmers	sum_control_kmers	p-value	case_ind_1	case_ind_2	case_ind_3	control_ind_1	control_ind_2	control_ind_3

Now you have to normalize and compare across species, since there are coverage differences between them - Example commands for 20X normalization:

    awk -F " " '{print $1,(($2/107483903941)*25000000000),(($3/100794792290)*25000000000)}' corrected_sums_case_out_w_bonf_sorted.txt >  normalized_corrected_case_w_bonf.txt
    awk '{if($3 == "0"){print;}}' normalized_corrected_case_w_bonf.txt > case_normalized_uniq.kmerDiff
    awk -F " " '($2 > 20){print;}'case_normalized_uniq.kmerDiff> case_normalized_uniq20.kmerDiff

You can check the number of unique k-mers based on the normalization 

    wc -l case_normalized_uniq20.kmerDiff

Once you have done this for all species, you can then compare the shared unique k-mers between species and the sexes - example commands:
    
    cut -f1 -d " " case_normalized_uniq20.kmerDiff > bifurca_case_normalized_uniq20_cut.kmerDiff
    sort bifurca_case_normalized_uniq20_cut.kmerDiff > bifurca_case_normalized_uniq20_cut_sorted.kmerDiff
    comm -12 picta_case_normalized_uniq20_cut_sorted.kmerDiff bifurca_case_normalized_uniq20_cut_sorted.kmerDiff > shared_PicBif_Male.txt



